src/hotspot/share/utilities/lockFreeQueue.hpp

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright (c) 2021, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#ifndef SHARE_UTILITIES_LOCKFREEQUEUE_HPP
+#define SHARE_UTILITIES_LOCKFREEQUEUE_HPP
+
+#include "memory/padded.hpp"
+#include "utilities/globalDefinitions.hpp"
+#include "utilities/pair.hpp"
+
+// Return status of a LockFreeQueue::try_pop() call.
+// See description for try_pop() below.
+enum class LockFreeQueuePopStatus {
+  success,
+  lost_race,
+  operation_in_progress
+};
+
+// The LockFreeQueue template provides a lock-free FIFO. Its structure
+// and usage is similar to LockFreeStack. It has inner paddings, and
+// provides a try_pop() function for the client to implement pop()
+// according to its need (e.g., whether or not to retry or prevent
+// ABA problem).
+//
+// \tparam T is the class of the elements in the queue.
+//
+// \tparam next_ptr is a function pointer.  Applying this function to
+// an object of type T must return a pointer to the list entry member
+// of the object associated with the LockFreeQueue type.
+template<typename T, T* volatile* (*next_ptr)(T&)>
+class LockFreeQueue {
+  T* volatile _head;
+  DEFINE_PAD_MINUS_SIZE(1, DEFAULT_CACHE_LINE_SIZE, sizeof(T*));
+  T* volatile _tail;
+  DEFINE_PAD_MINUS_SIZE(2, DEFAULT_CACHE_LINE_SIZE, sizeof(T*));
```

Comment:
LockFreeQueue should probably document it's padding.  And maybe should not have the second (trailing) padding, but leave that to the specific usage.  The trailing padding in the original was consistent with the usage there.  Other uses might not need any trailing padding, or might need less.

Reply:
Moved the padding.